clients/src/main/java/org/apache/kafka/clients/MetadataCache.java

```diff
@@ -0,0 +1,126 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients;
+
+import org.apache.kafka.common.Cluster;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.TopicPartition;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An internal mutable cache of nodes, topics, and partitions in the Kafka cluster. This keeps an up-to-date Cluster
+ * instance which is optimized for read access.
+ */
+class MetadataCache {
+    private final String clusterId;
+    private final List<Node> nodes;
+    private final Set<String> unauthorizedTopics;
+    private final Set<String> invalidTopics;
+    private final Set<String> internalTopics;
+    private final Node controller;
+    private final Map<TopicPartition, PartitionInfo> partitionsByTopicPartition;
+
+    private Cluster clusterInstance;
+
+    MetadataCache(String clusterId,
+                  List<Node> nodes,
+                  Collection<PartitionInfo> partitions,
+                  Set<String> unauthorizedTopics,
+                  Set<String> invalidTopics,
+                  Set<String> internalTopics,
+                  Node controller) {
+        this(clusterId, nodes, partitions, unauthorizedTopics, invalidTopics, internalTopics, controller, null);
+    }
+
+    MetadataCache(String clusterId,
+                  List<Node> nodes,
+                  Collection<PartitionInfo> partitions,
+                  Set<String> unauthorizedTopics,
+                  Set<String> invalidTopics,
+                  Set<String> internalTopics,
+                  Node controller,
+                  Cluster clusterInstance) {
+        this.clusterId = clusterId;
+        this.nodes = nodes;
+        this.unauthorizedTopics = unauthorizedTopics;
+        this.invalidTopics = invalidTopics;
+        this.internalTopics = internalTopics;
+        this.controller = controller;
+
+        this.partitionsByTopicPartition = new HashMap<>(partitions.size());
+        for (PartitionInfo p : partitions) {
+            this.partitionsByTopicPartition.put(new TopicPartition(p.topic(), p.partition()), p);
+        }
+
+        if (clusterInstance == null) {
+            updateClusterView();
+        } else {
+            this.clusterInstance = clusterInstance;
+        }
+    }
+
+    Cluster cluster() {
+        if (clusterInstance == null) {
+            throw new IllegalStateException("Cached Cluster instance should not be null, but was.");
+        } else {
+            return clusterInstance;
+        }
+    }
+
+    synchronized boolean removePartition(TopicPartition topicPartition) {
```

Comment:
Hmm.. If we remove a single partition, then won't that break `Cluster. partitionCountForTopic`? I think it might be safer to leave the stale data inside `Cluster`, but expose a new internal API to get the latest leader information only if it is not stale. For example, we could create a `PartitionInfoAndEpoch` class which we retain for each partition in `MetadataCache`. Then `Metadata` could expose an api which returns the leader info only if the cached epoch matches the epoch in `lastSeenLeaderEpochs`. What do you think?

Reply:
Well on the very next line we recompute the Cluster completely, so the partition count should always be valid for a given Cluster instance (it might be out of date if you've held on for Cluster for too long).

Reply:
My point is that the recomputed Cluster is incomplete. We use `partitionCountForTopic` in order to tell how many partitions a topic has, but it would no longer be an accurate count. Or am I missing something?

Reply:
Ah, sorry I misunderstood -- indeed we are forgetting about metadata for partitions which still exist. I'll try your suggestion